bin/run-template.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * Run a template command.
+ * Examples:
+ *    bin/run-template vendor/bin/parallel-lint DIR --
+ *    bin/run-template vendor/bin/parallel-lint DIR -- all
+ *    bin/run-template vendor/bin/parallel-lint DIR -- packages/mediawiki-api-base
+ */
+
+$splitKey = array_search('--', $argv);
+$templateCommand = implode( ' ', array_slice($argv,1,$splitKey-1) );
+$directory = implode( ' ', array_slice($argv,$splitKey+1,1) );
+
+if($directory === '') {
+    $directory = 'all';
+} elseif(count($argv) !== $splitKey +2) {
+    die("You're using it wrong!");
+}
+
+$dirs = [];
+if($directory === 'all') {
+    $dirs = array_merge(
+        array_filter(glob(__DIR__ . '/../packages/*'), 'is_dir'),
+        array_filter(glob(__DIR__ . '/../packages-dev/*'), 'is_dir')
+    );
+    foreach( $dirs as $dir ) {
+        $relativeDir = str_replace(realpath(__DIR__."/../"),'.',realpath($dir));
+        $dirs[] = $relativeDir;
+    }
+} else {
+    $dirs[] = $directory;
+}
+
+$finalExitCode = 0;
+foreach( $dirs as $dir ) {
+    $command = str_replace('DIR', $dir, $templateCommand);
+    echo "\033[0;32m" . "Running: " . "\033[0;36m" . $command . "\033[0m" . PHP_EOL;
+    $exitCode = runAndStreamCommand( $command, realpath(__DIR__ . '/../') );
+    echo "\033[0;32m" . "Exit code: " . "\033[0;36m" . $exitCode . "\033[0m" . PHP_EOL;
+    if($exitCode != 0) {
+        $finalExitCode = $exitCode;
+    }
+}
+
+exit( $finalExitCode );
+
+function runAndStreamCommand( $cmd, $cwd ) {
+    $descriptorspec = array(
+        0 => array("pipe", "r"),   // stdin is a pipe that the child will read from
+        1 => array("pipe", "w"),   // stdout is a pipe that the child will write to
+        2 => array("pipe", "w")    // stderr is a pipe that the child will write to
+     );
+     flush();
+     $process = proc_open($cmd, $descriptorspec, $pipes, $cwd, array());
+     if (is_resource($process)) {
+         while ( ( $c = fgetc($pipes[1]) ) !== false ) {
+             echo $c;
+             flush();
+         }
+     }
+     return proc_get_status($process)['exitcode'];
+}
```

Comment:
Personally, this look a bit overkill to run CI locally.
In my experience it's better to CI workflow setup and let it do the hard work. Locally run only the neccesary commands.

E.g like this: https://github.com/rectorphp/rector/blob/9c26ebf430a76d1dc4a65d3dc970705451c9b8fd/.github/workflows/weekly_pull_requests.yaml#L25-L27

Reply:
That is sneaky and interesting and something that I might have to look into!
Some of the non fix code sniffs in mediawiki-codesniffer are aimed at making sure things are correctly documented.
But having it as a tool that can help you, but not necessarily fail CI and just auto fix every week sounds quite appealing!